drivers: ethernet: add support for Xilinx GEM controller

ibirnbaum opened the pull request:
Add support for the Xilinx GEM Ethernet controller, which is integrated
in both the Xilinx Zynq and ZynqMP (UltraScale) SoC families. The driver
supports the management of a PHY attached to the respective GEM's MDIO
interface.

This driver was developed with ultimately the Zynq-7000 series in mind,
but at the time being, it is limited to use in conjunction with the
ZynqMP RPU (Cortex-R5) cores. The differences are minor when it comes
to the adjustment of the TX clock frequency derived from the current
link speed reported by the PHY, but for use in conjunction with the
Zynq-7000, some larger adjustments will have to be made when it comes
to the placement of the DMA memory area, as this involves the confi-
guration of the MMU in Cortex-A CPUs.

The driver was developed under the qemu_cortex_r5 target. The Marvell
88E1111 PHY simulated by QEMU is supported by the driver.

Limitations currently exist when it comes to timestamping or VLAN
support and other minor things. Those haven't been implemented yet,
although they are supported by the hardware. In order to be fully
supported by the ZynqMP APU, 64-bit DMA address descriptor format
support will be added.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum committed:
drivers: ethernet: add support for Xilinx GEM controller

Add support for the Xilinx GEM Ethernet controller, which is integrated
in both the Xilinx Zynq and ZynqMP (UltraScale) SoC families. The driver
supports the management of a PHY attached to the respective GEM's MDIO
interface.

This driver was developed with ultimately the Zynq-7000 series in mind,
but at the time being, it is limited to use in conjunction with the
ZynqMP RPU (Cortex-R5) cores. The differences are minor when it comes
to the adjustment of the TX clock frequency derived from the current
link speed reported by the PHY, but for use in conjunction with the
Zynq-7000, some larger adjustments will have to be made when it comes
to the placement of the DMA memory area, as this involves the confi-
guration of the MMU in Cortex-A CPUs.

The driver was developed under the qemu_cortex_r5 target. The Marvell
88E1111 PHY simulated by QEMU is supported by the driver.

Limitations currently exist when it comes to timestamping or VLAN
support and other minor things. Those haven't been implemented yet,
although they are supported by the hardware. In order to be fully
supported by the ZynqMP APU, 64-bit DMA address descriptor format
support will be added.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

jukkar commented on drivers/ethernet/Kconfig.xlnx_gem:
```diff
@@ -0,0 +1,1387 @@
+
+# ---------- GEM0 ----------
+
+config ETH_XLNX_GEM_PORT_0
```
There are lot of "same" options for the supported ports. Could we move all these settings to DTS in order to avoid this duplication?

ibirnbaum replied:
Hi Jukka, thanks for the quick review, just finished round one of refactoring based on the results of the automated code analysis. From my point of view, all the properties of the controller that can currently be tweaked using menuconfig should work just as well when supplied via the device tree. I just wasn't sure to what degree highly customized bindings are welcome, and if there are any guidelines for identifiers. E.g., adding the well known property "clock-frequency" to one particular device of a class that doesn't have this property by default which requires it probably isn't such a big deal. But for all those little things that can be tweaked, I'd be concerned about re-usability / common usefulness of the corresponding property names. If this is not an issue, I'd give moving the entire configuration to the DT a try. I'm not entirely sure but I doubt that, while improved readability of the code is surely given, this change would eventually eliminate the need to have separate mentions of the maximum of 4 available device instances (in the UltraScale, the Zynq only has GEM0/1) both in the driver's header and the source file. I've tried this before and at some point ran into the situation that I would have needed recursive macros.

ibirnbaum replied:
I've thought about this issue again and one thing came to my mind: there's at least one dependency that is automatically implemented when using Kconfig that I'm not sure of whether it can be expressed when configuring features via the DT or not. The controller supports hardware offloading of checksum generation while transmitting and checksum verification while receiving. Using this feature is generally desireable, so it's set to default y for both directions. The thing is: the QEMU simulation of the GEM, which does have a few quirks anyways, just doesn't implement this. Enable the offloading and you'll see ARP packets going in and out of QEMU, but everything that's checksummed is not going anywhere. Therefore, the checksum offloading items currently have a dependency in Kconfig that is "depends on !QEMU_TARGET". If such a dependency can be expressed in the device tree data, then it should be possible to move the configuration from Kconfig to DT.

jukkar commented on drivers/ethernet/Kconfig.xlnx_gem:
```diff
@@ -0,0 +1,1387 @@
+config ETH_XLNX_GEM_PORT_0_DISC_RX_AHB_UNAVAIL
+	bool "Discard RX packets when no AHB resource is available"
+	depends on ETH_XLNX_GEM_PORT_0
+	default n
```
All the options are disabled by default, so there is no need to mention this line. Could you remove this one and all the other `default n` lines below.

ibirnbaum replied:
Just removed all the "default n" lines while fixing copy&paste mistakes both in the driver sources and the KConfig relating to two configuration items per controller.

jukkar commented on drivers/ethernet/phy_xlnx_gem.h:
```diff
@@ -0,0 +1,119 @@
+/*
```
The #34718 is pending and is providing MDIO and PHY driver support. Is it possible to utilize that generic support here?

ibirnbaum replied:
I'll have a look into that tomorrow. At first glance it seemed to me that up until now, all the drivers supporting PHY management all did their own thing. If that interface provides link state and link speed information and is suitable for interrupt-driven operation where the interrupt line is actually that of the controller itself plus a corresponding bit set in the interrupt status register, plus it allows the integration of PHY types not previously supported, I guess that might work.

ibirnbaum replied:
I've had a first look at this PR and at the time being, it doesn't look suitable:
- It's unclear how interrupt handling is going to work. If this is supposed to be implemented by using a dedicated ISR attached to an interrupt line specified by an interrupts property within the PHY's device tree entry, as mentioned above, that's incompatible with the way the GEM signals the PHY interrupt.
- Gigabit is not considered.
- The API doesn't seem to have any method of specifying what shall actually be advertised to the link partner. The current generic MII driver implementation triggers auto-negotiation upon the link going up, looks at the link partner capability register, and returns the best possible link from that set of data. I need to be able to tell, for example, a Gigabit PHY to only advertise 100M FDX or configurations like "100M FDX and anything below that" even if the link partner also has GBit capability.
- The PHY driver implementation to be used in a certain target is specified via the PHY's "compatible" property in the device tree. I personally need auto-detection. We're a large scale device manufacturer, and our PCB manufacturer was given a list of pin-compatible, tested PHYs that is updated whenever a part on that list reaches EOL status. For each production run, they'll use whatever PHY is currently in their parts bin. Therefore, two devices that look identical on the outside may have anything between subtle differences within a PHY register all the way to a different register layout.
- There's also one particular corner case this implementation doesn't consider - but neither does mine yet: Each GEM, in both the Zynq and the UltraScale, has its own MDIO bus, which is accessed via the respective GEM's register space. Yet, we have a device which uses GEM0 & GEM1, each connected to its own physical Ethernet port. Yet, there just weren't enough free I/O pins in order to expose the MDIO interface of GEM1. Therefore, both PHYs are attached to the MDIO bus of GEM0. Which means that not only does each GEM instance need to know the address of its respective PHY on the MDIO bus, but also the address of the base device via which it can be accessed. Considering that the device driver instances of both GEMs might want to talk to their respective PHYs simultaneously, this scenario will eventually require a mutex concept for shared MDIO access.

ibirnbaum replied:
Another feature the suggested PHY framework is currently missing: LED configuration.

ibirnbaum committed:
dts: arm: xilinx: zynqmp: add device tree data for GEM Ethernet controllers

Add the device tree data for the 4 Ethernet controllers integrated into
the ZynqMP SoC, GEM0 to GEM3.

Signed-off-by: Immo Birnbaum <Immo.Birnbaum@weidmueller.com>

ibirnbaum commented:
The configuration has been changed to all-DT based, the DT data for the ZynqMP has been added as a 2nd commit.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1724 @@
+	/* Precondition checks using assertions */
+
+	/* Valid PHY address and polling interval, if PHY is to be managed */
+	if (dev_conf->init_phy == 1) {
```
I just wonder if it would be more clearer if `init_phy` is made boolean as it seems to be either 0 or 1. Then the reader would not need to guess if the integer value has some special meaning, and we could write here
`if (dev_conf->init_phy) {`

ibirnbaum replied:
All device configuration parameters of type uint8_t with boolean semantics have been changed to type bool. All if-statements evaluating them have been changed to the syntax you suggested.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1724 @@
+	 */
+	k_work_init(&dev_data->tx_done_work, eth_xlnx_gem_tx_done_work);
+	k_work_init(&dev_data->rx_pend_work, eth_xlnx_gem_rx_pending_work);
+	k_delayed_work_init(&dev_data->phy_poll_delayed_work,
```
This work API is deprecated and replaced by `k_work_init_delayable`, this needs to be converted before the driver can be merged.

ibirnbaum replied:
The delayable work API calls have been updated to the current version.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1724 @@
+	dev_conf->config_func(dev);
+
+	/* Submit initial PHY status polling delayed work */
+	k_delayed_work_submit(&dev_data->phy_poll_delayed_work, K_NO_WAIT);
```
same here

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1724 @@
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+	uint32_t reg_val = 0;
+
+	if (dev_data->started == 1) {
```
Boolean variable would be more clearer here.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1724 @@
+		caps |= ETHERNET_LINK_10BASE_T;
+	}
+
+	if (dev_conf->enable_rx_chksum_offload == 1) {
```
These vars, that are "iffed" here and below, seem to be boolean by nature. You could even combine all the booleans in dev_conf to bit fields to save some space.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1724 @@
+	detect_rc = phy_xlnx_gem_detect(dev);
+
+	if (detect_rc == 0 &&
+		dev_data->phy_id != 0x00000000 &&
+		dev_data->phy_id != 0xFFFFFFFF &&
+		dev_data->phy_access_api != NULL) {
```
Minor nit, if you indent these one or more steps right, then it is easier to read (because the block below is not in the same level).

ibirnbaum replied:
I've adjusted the indentation for this code block.

jukkar commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -0,0 +1,743 @@
+	uint8_t				phy_advertise_lower;
+	uint32_t			phy_poll_interval;
+	uint8_t				defer_rxp_to_queue;
+	uint8_t				defer_txd_to_queue;
+
+	enum eth_xlnx_amba_dbus_width	amba_dbus_width;
+	enum eth_xlnx_ahb_burst_length	ahb_burst_length;
+	enum eth_xlnx_hwrx_buffer_size	hw_rx_buffer_size;
+	uint8_t				hw_rx_buffer_offset;
+
+	uint8_t				rxbd_count;
+	uint8_t				txbd_count;
+	uint16_t			rx_buffer_size;
+	uint16_t			tx_buffer_size;
+
+	uint8_t				ignore_ipg_rxer;
+	uint8_t				disable_reject_nsp;
+	uint8_t				enable_ipg_stretch;
+	uint8_t				enable_sgmii_mode;
+	uint8_t				disable_reject_fcs_crc_errors;
+	uint8_t				enable_rx_halfdup_while_tx;
+	uint8_t				enable_rx_chksum_offload;
+	uint8_t				disable_pause_copy;
+	uint8_t				discard_rx_fcs;
+	uint8_t				discard_rx_length_errors;
+	uint8_t				enable_pause;
+	uint8_t				enable_tbi;
+	uint8_t				ext_addr_match;
+	uint8_t				enable_1536_frames;
+	uint8_t				enable_ucast_hash;
+	uint8_t				enable_mcast_hash;
+	uint8_t				disable_bcast;
+	uint8_t				copy_all_frames;
+	uint8_t				discard_non_vlan;
+	uint8_t				enable_fdx;
+	uint8_t				disc_rx_ahb_unavail;
+	uint8_t				enable_tx_chksum_offload;
+	uint8_t				tx_buffer_size_full;
+	uint8_t				enable_ahb_packet_endian_swap;
+	uint8_t				enable_ahb_md_endian_swap;
```
Many of these uint8_t look like boolean values, so if we combine them to bool bitfield, then we could save some mem.

ibirnbaum replied:
The entire block below tx_buffer_size, now typed bool instead of uint8_t, is now a bitfield.

jukkar reviewed (commented):
There was lot of code so I probably missed things. Anyway, looks good, just some minor comments from me.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+	k_work_init(&dev_data->tx_done_work, eth_xlnx_gem_tx_done_work);
+	k_work_init(&dev_data->rx_pend_work, eth_xlnx_gem_rx_pending_work);
+	k_work_init_delayable(&dev_data->phy_poll_delayed_work,
+		eth_xlnx_gem_poll_phy);
```
Indentation

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+	 */
+	if ((reg_val & ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT) != 0) {
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
```
indentation

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_ISR_OFFSET);
```
ditto

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+	}
+	if ((reg_val & ETH_XLNX_GEM_IXR_FRAME_RX_BIT) != 0) {
+		sys_write32(ETH_XLNX_GEM_IXR_FRAME_RX_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
```
ditto

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+		sys_write32(ETH_XLNX_GEM_IXR_FRAME_RX_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
+		sys_write32(ETH_XLNX_GEM_IXR_FRAME_RX_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_ISR_OFFSET);
```
ditto

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+		k_sem_take(&(dev_data->txbd_ring.ring_sem), K_FOREVER);
+	} else {
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
```
indentation

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+			k_sem_give(&(dev_data->txbd_ring.ring_sem));
+		} else {
+			sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+				dev_conf->base_addr + ETH_XLNX_GEM_IER_OFFSET);
```
here too

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+		k_sem_give(&(dev_data->txbd_ring.ring_sem));
+	} else {
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IER_OFFSET);
```
and here. Actually all sys_writexx() calls seem to have these indentation issue, could you check them all in one go.

jukkar commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1733 @@
+#endif
+
+	/* Block until TX has completed */
+	sem_status = k_sem_take(&dev_data->tx_done_sem, K_MSEC(5000));
```
The timeout is quite long, is it really plausible that we could wait this long. Could we make this timeout shorter?

jukkar reviewed (commented):
Looks good, just minor style nits

jukkar reviewed (approved):
LGTM

ibirnbaum commented:
I've had another go at formatting the code. The coding guidelines check didn't really appreciate any attempts to split the lengthy "(uint32_t)(&dev_data->txbd_ring.first_bd[curr_bd_idx].ctrl)" statements used to obtain the address of a buffer descriptor control/address word which take up the best part of a line. I hope everything looks as it should by now.

I've also reduced the timeout when waiting for the TX completion semaphore, the 5 seconds were a random pick in order to assure during development that something had really gone wrong regarding the transmission or the handling of the transmission completion indication whenever a timeout occurred (instead of a high load situation during which a lot of time might be spent processing RX data in the ISR).

bbolen commented on dts/bindings/ethernet/xlnx,gem.yaml:
```diff
@@ -0,0 +1,353 @@
+        item must be set to the clock frequency of the PLL supplying the
+        respective GEM's TX clock - by default, this is the IO PLL.
+
+    clk-ctrl-reg:
```
Can this be changed to use a clock driver instead of using a hardcoded address to another block?

mbolivar-nordic replied:
I agree, it looks like this should probably take a phandle to another node with compatible fixed-clock. cc @galak

ibirnbaum replied:
This whole clock-thing is a bit of a complicated issue.

In the Xilinx SoCs, all of the clock sources, frequencies, multipliers/dividers etc. relating to the Processor System, integrated peripherals and the programmable logic part have to be configured as part of the FPGA synthesis project in the Vivado IDE (which unfortunately results in redundant work, the whole clock config has to be specified first in Vivado and then in the device tree of the target board, with maybe some defaults at the SoC level of the DT matching the most common configurations). A by-product of the generated FPGA bitstream is an auto-generated C source file which contains loads of register read&write commands which effectively replicate those settings into what eventually becomes the first stage bootloader binary, built by the Xilinx SDK. Therefore, driver for Xilinx peripherals shouldn't mess with any clock configurations during start up, as the system is always pre-configured when booting into the OS.

The only exception to that rule are input clock prescalers that can be found for example in the UART for the generation of the configured baud rate, CAN has such a prescaler as well, and in the Ethernet Controller, a prescaler of the input clock is required for the TX clock generation based on the link speed reported by the PHY. The peripheral drivers need not know or actively alter the setting which PLL drives their respective device, they just need to know the actual input frequency, regardless of the clock source, based on which the respective prescaler is calculated, and this value has to be written to the respective peripheral's clock configuration register without modifying any other contents of that register. Therefore, the additional "clock-frequency" item in the DT binding is sufficient for the calculation part and is consistent with how input clocks are specified for other device classes such as UARTs, timers etc. As long as there's no "reg" property contained in the "fixed-clock" class (see explanation below), I don't think that there's any real benefit in referencing a separate clock node in the DT for the specification of the plain source clock frequency.

Regarding the clock configuration register to which the prescaler eventually has to be written, there's a bit of an unfortunate design decision by Xilinx: the peripheral clock configuration register are one (Zynq, SLCR) or three (UltraScale, CRF_APB, CRL_APB, IOU_SLCR) separate register spaces, so that each peripheral's clock configuration register is not contained within the register space specified by each peripheral's "reg" property. My initial idea on how to solve this issue was to create one/three clock configuration node(s) in the DT, based on some existing nodes of such type I think I found in the STM32 device trees. This would have allowed to specify (most likely at the board level) the HW oscillator frequency and all the routing / multiplier / divider / clocking scheme informations that configure the entire system's clocks. I added the calculation of frequencies derived from those base settings in the SoC init code, and there was getters for everything at that level, too. Setters were provided for each peripheral's prescaler where required, without the driver having to know its associated register address (this was resolved via the peripheral's base address). The only redundancy (aside from having to set up everything in Vivado first, then having to copy all the data from the synthesis report to the board's DT) then would have been SYS_CLOCK_HW_CYCLES_PER_SEC. The GEM driver in development, and also future implementations such as a CAN, I2C, ... driver, could have just called a getter to obtain the respective input clock frequency, could have calculated the prescaler, and could have called a setter with the device's base address and the result of the calculation as parameter. That way, no knowledge of bit masking and shifting in the config register (differs slightly between Zynq and UltraScale here and there) would have been required either, one #ifdef less. The moment when all this fell apart was when I realized that even if I shifted the existing UART driver over to this scheme, UART devices always require a "clock-frequency" value. At that point, there would have been an inconsistency, so I threw it all out again and opted to make it all consistent again by having the GEM have a "clock-frequency" property as well.

This still doesn't solve the control register issue, though. The register address could be determined within the driver, but that would require a nested switch first over the current SoC family, then over the current device's base address. Admittedly, that's not a great solution - as the driver currently is, the SoC family switch is already back in for writing to the clock control register due to the slightly different bit positions. So if a "fixed-clock" node can have a "reg" property without having to introduce an additional (if optional) property that's useless pretty much every other use case, then I'd agree that referencing a separate node that contains both the frequency and the register location would be the best way to solve the issue. But if that is not the case, I'm short on ideas on how to solve the issue elegantly.

mbolivar-nordic replied:
fixed-clock indeed has no reg; you can see this here:

https://docs.zephyrproject.org/latest/reference/devicetree/bindings/clock/fixed-clock.html#dtbinding-fixed-clock

I'd rather not block this PR on that detail then. It seems like this is a pretty special case, and your users are going to have to jump through enough hoops as it is.

However, while I was looking at the bindings index, I realized that the binding added for this compatible doesn't follow the usual practice for multiline descriptions, which is to use:

```yaml
  description: |
    My very long description
    goes here
```

This is a bit of a game of whack-a-mole to keep clean, but I've just done another pass while I was here: https://github.com/zephyrproject-rtos/zephyr/pull/36189

Would you mind following suit?

ibirnbaum replied:
I'll change the multiline formatting, no problem!

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+			base_addr, phy_addr, reg_addr);
+		return 0;
+	}
+	poll_cnt = 0;
```
i'd move this down to the do-while below, just in case code gets added between here and there in the future.

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+			base_addr, phy_addr, reg_addr);
+		return;
+	}
+	poll_cnt = 0;
```
move down to do-while loop

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+		PHY_MRVL_COPPER_PAGE_SWITCH_REGISTER,
+		PHY_MRVL_GENERAL_CONTROL_1_PAGE);
+
+	phy_data = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
```
is reading the page register back necessary?

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+			dev->name, dev_data->phy_addr);
+		return;
+	}
+	retries = 0;
```
move to the while loop below where it is used

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+	 * 0, 2, 3, 5, 7. Reset is performed immediately, bit [15] is
+	 * self-clearing.
+	 */
+	phy_data  = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
```
can this be reworked so that you can use the alaska_reset() function above instead of open coding this?

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+	 * 0, 2, 3, 5, 7. Reset is performed immediately, bit [15] is self-
+	 * clearing.
+	 */
+	phy_data  = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
```
same comment about the reset.  maybe pass autoneg as a parameter to the reset function?

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+
+	phy_data = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
+		dev_data->phy_addr, PHY_TI_REG_BMCR);
+	phy_data |= 0x8000; /* Reset bit */
```
0x8000 should be a macro like the code for the Marvell phy

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+	if (dev_conf->enable_fdx) {
+		if (dev_conf->max_link_speed == LINK_100MBIT) {
+			/* Advertise 100BASE-TX, full duplex */
+			phy_data |= (1 << 8);
```
all of these bits should be macros as well.  Pretty much throughout this whole function there are open coded bits that should probably be macros.

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+	phy_data = phy_xlnx_gem_mdio_read(dev_conf->base_addr,
+		dev_data->phy_addr, PHY_TI_REG_MISR1);
+
+	if ((phy_data & 0x400) != 0) {
```
macros for 0x400, etc

bbolen commented on drivers/ethernet/phy_xlnx_gem.c:
```diff
@@ -0,0 +1,975 @@
+			 * in the device's run-time data struct.
+			 */
+			for (list_iter = 0; list_iter <
+				(sizeof(phy_xlnx_gem_supported_devs) /
```
ARRAY_SIZE(phy_xlnx_gem_supported_devs) from sys/util.h

bbolen commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1679 @@
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+/* IRQ configuration function for GEM0 */
+ETH_XLNX_GEM_CONFIG_IRQ_FUNC(0);
```
You could declare a super macro that expands to all of these, i.e.
ETH_XLNX_GEN_DEV(0) which would expand to the 5 macros below

bbolen commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -0,0 +1,1679 @@
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+
+	uint16_t tx_data_length = 0;
```
In most of these functions, many of the local variables do not need to be assigned a value because they are assigned again before they are used.  For example, tx_data_length and tx_data_remaining

carlescufi commented:
@ibirnbaum can you follow-up to the review by @bbolen please?

ibirnbaum commented:
I'll take care of the issues mentioned above, might take a day or two though.

ibirnbaum commented:
I've modified the code based on the comments from the previous review, I hope I haven't overlooked anything:
- Changed the multi-line formatting in the DT binding YAML as suggested by Marti.
- Removed the clk-ctrl-reg property and integrated as a 2nd entry in the reg property of each controller instance. Extraction from the device tree header into the device config struct was modified accordingly.
- Super-macroed the device initialization.
- Moved initialization of iterator variables closer to the loops in which they are used.
- Removed explicit initialization to zero of all variables whose initial value is assigned by a hardware read, device config / run-time data values, for-loop etc. Explicit initialization only remains for variables where the contents of a register or the capability flags are OR-ed into a variable that should initially be zero.
- The repeated resets of the Marvell Alaska PHYs triggered in the config function are now implemented by calling the reset function. Masking out of the auto-neg enable bit now takes place as the first step of the config function. This bit can be masked out individually, there's no dependency so that it has to be set in the same write access as the reset bit.
- TI PHY registers, bit identifiers, bit masks etc. are now based on macros, as with the Marvell PHYs.
- PHY detection function now uses ARRAY_SIZE to set up the for-loop iterating over the supported PHYs list.
- Removed one unnecessary PHY register readback in the Marvell config function. There is a case where a register has to be read twice in a row for the TI PHYs, this has been documented.

mbolivar-nordic reviewed (approved):
Thanks for taking the time to write such nice binding documentation and fixing the multiline strings.

ibirnbaum commented:
Fixed one minor glitch in the bindings file - a parameter which originally was optional and then became mandatory was still referred to as being optional in the description text.
